File: articles/calculus-for-systems-modeling/articles/introduction-to-partial-differential-equations/c/pde_diffusion_grid_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main(void){
  const int grid_points = 51;
  const int steps = 100;
  const double diffusivity = 0.1, dx = 1.0, dt = 0.25;
  const double ratio = diffusivity * dt / (dx * dx);
  double field[grid_points];
  double updated[grid_points];

  for(int i=0; i<grid_points; i++){ field[i] = 0.0; }
  field[grid_points/2] = 1.0;

  printf("step,time,center_value,total_mass,max_value,min_value,stability_ratio,warning\n");
  for(int step=0; step<=steps; step++){
    double total = 0.0, maxv = field[0], minv = field[0];
    for(int i=0; i<grid_points; i++){
      total += field[i] * dx;
      if(field[i] > maxv) maxv = field[i];
      if(field[i] < minv) minv = field[i];
    }
    printf("%d,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,Explicit diffusion schemes require stability checks boundary and grid assumptions shape results.\n",
      step, step*dt, field[grid_points/2], total, maxv, minv, ratio);

    for(int i=0; i<grid_points; i++){ updated[i] = field[i]; }
    for(int i=1; i<grid_points-1; i++){ updated[i] = field[i] + ratio*(field[i+1] - 2*field[i] + field[i-1]); }
    updated[0] = 0.0;
    updated[grid_points-1] = 0.0;
    for(int i=0; i<grid_points; i++){ field[i] = updated[i]; }
  }
  return EXIT_SUCCESS;
}

File: articles/calculus-for-systems-modeling/articles/introduction-to-partial-differential-equations/julia/pde_diffusion_grid_audit.jl
function initialize_field(grid_points)
    field = zeros(Float64, grid_points)
    field[cld(grid_points, 2)] = 1.0
    return field
end

function diffusion_step(field, ratio)
    updated = copy(field)
    for i in 2:(length(field)-1)
        updated[i] = field[i] + ratio * (field[i+1] - 2 * field[i] + field[i-1])
    end
    updated[1] = 0.0
    updated[end] = 0.0
    return updated
end

grid_points = 51
diffusivity = 0.1
dx = 1.0
dt = 0.25
steps = 100
ratio = diffusivity * dt / dx^2
field = initialize_field(grid_points)

println("step,time,center_value,total_mass,max_value,min_value,stability_ratio,warning")
for step in 0:steps
    time = step * dt
    center_value = field[cld(grid_points, 2)]
    total_mass = sum(field) * dx
    println(join((step, time, center_value, total_mass, maximum(field), minimum(field), ratio, "Explicit diffusion schemes require stability checks boundary and grid assumptions shape results."), ","))
    global field = diffusion_step(field, ratio)
end
